Feature/inthad 157 extend the system api to receive tuples of data

apattnaik0721013 opened the pull request:
Generalize different sorts of slam inputs for the orb-slam2.
 
1. boost::iterator_adaptor is used for generic data_source traversing.
2. factory pattern is used for the specialized data_source.
3. SemanticInfo.h is moved to "ext" folder.

Note: currently forward iterator is used which will support only next() to get next input. it can be change to different iterator with little change based on the requirement.
also the I added the iterator to the generic (base) data_source which can be used for all sorts of different data_source.

paul-michalik commented on ext/SemanticInfo.h:
```diff
@@ -0,0 +1,75 @@
+		boost::optional<pos_info> pos;
+	};
+
+#define GET_IMG(sensor_data) std::get<0>(sensor_data)
```
No macros! Please remove the macros and never use macros in C++ code again. You can either (1) retrieve the data via the typed version of `get` (= `get<cv::Mat>`, `get<tsr_info>` etc.), (2) define specific access functions `cv::Mat const& get_image` etc.. (3) not use `std::tuple` if that is a problem.

paul-michalik commented on Examples/Monocular/semantic_monocular.hpp:
```diff
@@ -0,0 +1,213 @@
+/**
```
This isn't required. Remove this disclaimer since this is **NOT** part of ORB-SLAM2.

paul-michalik commented on Examples/Monocular/semantic_monocular.hpp:
```diff
@@ -0,0 +1,213 @@
+* You should have received a copy of the GNU General Public License
+* along with ORB-SLAM2. If not, see <http://www.gnu.org/licenses/>.
+*/
+
```
`#pragma once`

paul-michalik commented on Examples/Monocular/semantic_monocular.hpp:
```diff
@@ -0,0 +1,213 @@
+#include <stdexcept>
+# include <boost/iterator/iterator_adaptor.hpp>
+
+# ifndef BOOST_NO_SFINAE
```
Use `<type_traits>`! And for all type trait related stuff use `std::enable_if` etc.

paul-michalik commented on Examples/Monocular/semantic_monocular.hpp:
```diff
@@ -0,0 +1,213 @@
+	template <class OtherValue>
+	data_source_iter(
+		data_source_iter<OtherValue> const& other
+# ifndef BOOST_NO_SFINAE
```
`std::enable_if` and remove conditional compilation.

paul-michalik commented on Examples/Monocular/semantic_monocular.hpp:
```diff
@@ -0,0 +1,213 @@
+	)
+		: super_t(other.base()) {}
+
+# if !BOOST_WORKAROUND(__GNUC__, == 2)
```
What? I need to test it, but this has to work... Will check later.

paul-michalik commented on include/System.h:
```diff
@@ -81,11 +81,16 @@ class System
+    // Process the given monocular frame
+    // Input images: RGB (CV_8UC3) or grayscale (CV_8U). RGB is converted to grayscale.
+    // Returns the camera pose (empty if tracking fails).
+    cv::Mat TrackMonocular(std::tuple<ext::image_t, ext::time_point_t, ext::sensor_info> const&);
```
What if we alias as follows:
```
using ext::tsr_info_opt_t = boost::optional<std::vector<traffic_sign>>;
using ext::pos_info_opt_t = boost::optional<ext::pos_info>;
```
Then the API becomes:
```
cv::Mat TrackMonocular(std::tuple<ext_time_point_t, ext::image_t, ext::tsr_info_opt_t, ext::pos_info_opt_t> const& msg_);
```

apattnaik0721013 commented:
@paul-michalik 
1. currently Jason file contain frameid and std::vector<traffic_sign>.
2. semantic_data_source ::next() will return a map <frameid , std::vector<traffic_sign> >
   Though I can take framid as argument and return std::vector<traffic_sign>, but it will move to more specialization code.
3. run_slam_loop() ,
     a.  I call semantic_data_source ::next() and store the return map value as  traffic_signs_map
     b.  check if current framid is same as traffic_signs_map-> framid, the I use std::vector<traffic_sign>
     c.  if current framid is not same as traffic_signs_map->framid,  I wont call semantic_data_source ::next()

The point to consider:
1. from ORB-SLAM2 code , I remove the map <frameid , std::vector<traffic_sign> > usage instead I used tsr_info_opt_t data which is boost::optional<std::vector<traffic_sign>>
2. but in semantic_moncular.cc , I used the map <frameid , std::vector<traffic_sign> > which will allow me to verify the particular trafficsign frame_id with current frame_id.

I guess it wont create any problem and also to avoid use of map , we need to verify the trafficsign frame_id with the current frame_id , which we can do by passing frame_id to the semantic_data_source.
in this case we might have to compromise with generic next() iterator.

paul-michalik commented on ext/SemanticInfo.h:
```diff
@@ -46,9 +24,16 @@ namespace ORB_SLAM2 {
+        using tsr_info_opt_t = boost::optional<std::vector<traffic_sign>>;
+        using pos_info_opt_t = boost::optional<pos_info>;
+
         struct sensor_info {
```
Please remove this `sensor_info` struct. Let's stick with the original proposal, where we have a tuple of 4 readings without intermediate structures: `std::tuple<ext::time_point_t, ext::image_t, ext::tsr_info_opt_t, ext::pos_info_opt_t>`

paul-michalik commented on src/Frame.cc:
```diff
@@ -353,7 +353,7 @@ void Frame::AssignFeaturesToGrid()
  {
      if (std::get<ext::sensor_info>(_sensor_input).tsr) {
```
Please use `if` initialization `if (auto ts = std::get<ext::tsr_info_opt_t>(_sensor_input)) {...` and the dereferencing operator inside the `if` scope: `*ts`

apattnaik0721013 commented:
@paul-michalik 

I updated code as per your  your review.

paul-michalik commented on CMakeLists.txt:
```diff
@@ -39,28 +45,39 @@ ${Pangolin_INCLUDE_DIRS}
+    DESTINATION include/orb-slam2/ext 
+    FILES_MATCHING PATTERN *.h)
+
+file(GLOB_RECURSE source_list "src/*.cc" "src/*.cpp")
```
Remove the globbing expressions! Use the explicit file list from previous version.

paul-michalik commented on Examples/Monocular/semantic_monocular.cc:
```diff
@@ -110,41 +81,41 @@ class slam_object {
     }
 };
 
+
```
We need to change the format for the data container here. The data source should have the semantics of a stream of `std::tuple<time_point_t, ...>`. The reader will then retrieve each element by either `get_next` or, theoretically, by a streaming operator. Two possibilities: (1) We add a time stamp to the data and use the ID as key or (2) we use a time stamp as key. In all cases the ts extractor needs to be changed. Having real time stamps would also help in predicting speed between frames...

paul-michalik commented:
I know that looks fine, but please follow up on that...

apattnaik0721013 commented:
@paul-michalik 
rebased to master
I am going to address to use tuple in data source.

apattnaik0721013 commented:
@paul-michalik 
1. traffic sign has <timestamp,traffic_sign_map_t>
2. actual slam need std::tuple<ext::image_t, ext::time_point_t, ext::tsr_info_opt_t, ext::pos_info_opt_t>
3. Do you want me to implement as observer pattern to collect all the std::tuple<ext::image_t, ext::time_point_t, ext::tsr_info_opt_t, ext::pos_info_opt_t> from different data source (example-: from traffic sign and gps data)
In pictorial:
![image](https://user-images.githubusercontent.com/30471366/36904688-67436360-1e57-11e8-8fb9-c004acb63d8c.png)

Let me know if I understood correctly.

paul-michalik commented:
Yes, this looks good! Just make `time_point_t` the first item of the tuple. Besides of that, the implementation from `System::TrackMonocular` seems ok now. What we need to do is to define a more realistic format for the data in the "data source". The central sentence is this:

> The data source shall have the semantics of a stream of `std::tuple<time_point_t, ...>`.  

We don't have that right now, but we can simulate it. At the end we should be having a JSON object of the type
```
{
    "time_point":  000000,
    "image": 000000,
    "tsr_info":  [
         {
             "class_id": 27,
             "confidence": 0.8730757832527161,
             "box": [0.1, 0.2, 0.4, 0.4]
         },
         { ... }
    ],
    "pos_info": {
        "pos": [47.407614681869745, 8.553115781396627, 451.79998779296875],
        "cov": [0.1, 0.2, 0.3, 0.2, 0.1, 0.2]
    }
}
```

paul-michalik commented:
@apattnaik0721013 Please remove the rebasing commit! It overwrites the changes from this commit which fix the issues with CMake `custom_target`. I'll mark the important parts in the review...

paul-michalik commented on include/System.h:
```diff
@@ -81,11 +81,16 @@ class System
+    // Process the given monocular frame
+    // Input images: RGB (CV_8UC3) or grayscale (CV_8U). RGB is converted to grayscale.
+    // Returns the camera pose (empty if tracking fails).
+    cv::Mat TrackMonocular(std::tuple<ext::image_t, ext::time_point_t, ext::tsr_info_opt_t, ext::pos_info_opt_t> const&);
```
Please re-order the items as follows: `std::tuple<ext::time_point_t, ext::image_t, ext::tsr_info_opt_t, ext::pos_info_opt_t>`. Let's fix this as follows: a `time_point_t` is always a first item in a tuple, image is second.

paul-michalik commented on CMakeLists.txt:
```diff
@@ -45,39 +45,27 @@ ${Pangolin_INCLUDE_DIRS}
 add_subdirectory(Thirdparty/DBoW2)
 add_subdirectory(Thirdparty/g2o)
 
```
This needs to be included. You picked the wrong snippet while resolving the conflict.

paul-michalik commented on CMakeLists.txt:
```diff
@@ -86,10 +74,9 @@ if(CMAKE_COMPILER_IS_GNUCXX)
 endif()
 
 target_link_libraries(orb-slam2
```
The same here. This has to be included in the commit.

paul-michalik commented on CMakeLists.txt:
```diff
@@ -159,7 +145,13 @@ if(BUILD_EXPERIMENTS)
 endif()
 
-# Scripts and utilities as target
+add_custom_target(orb-slam2-ext SOURCES
```
This was replaced in this branch and needs to be removed.

paul-michalik commented on CMakeLists.txt:
```diff
@@ -125,15 +111,15 @@ if(BUILD_ALL_EXAMPLES)
 
     add_executable(semantic_monocular_video Examples/Monocular/semantic_monocular_video.cpp)
-    target_link_libraries(semantic_monocular_video orb-slam2 ${Boost_LIBRARIES})
+    target_link_libraries(semantic_monocular_video orb-slam2 filesystem)
```
Dito. Use the `${Boost_LIBRARIES}` variable until CMake updates `FindBoost.cmake` for Boost 1.66

paul-michalik commented on CMakeLists.txt:
```diff
@@ -45,39 +45,27 @@ ${Pangolin_INCLUDE_DIRS}
 add_subdirectory(Thirdparty/DBoW2)
 add_subdirectory(Thirdparty/g2o)
 
-# INTERFACE targets seems to be broken for "Visual Studio" generators. 
```
This has to be included

paul-michalik commented on CMakeLists.txt:
```diff
@@ -45,39 +45,27 @@ ${Pangolin_INCLUDE_DIRS}
-    DESTINATION include/orb-slam2/ext 
-    FILES_MATCHING PATTERN *.h)
-
-file(GLOB_RECURSE source_list "src/*.cc" "src/*.cpp")
```
This needs to go.

paul-michalik commented on CMakeLists.txt:
```diff
@@ -86,10 +74,9 @@ if(CMAKE_COMPILER_IS_GNUCXX)
 endif()
 
 target_link_libraries(orb-slam2
-    orb-slam2-ext
```
This need to be included

apattnaik0721013 commented:
@paul-michalik 
I have address all your comments and updates the source accordingly.

paul-michalik commented on ext/messages.h:
```diff
@@ -0,0 +1,36 @@
+        using image_t = cv::Mat;
+        using tsr_info_opt_t = boost::optional<std::vector<traffic_sign>>;
+        using pos_info_opt_t = boost::optional<pos_info>;
+        using slam_input_t = std::tuple<time_point_t,image_t,tsr_info_opt_t, pos_info_opt_t>;
```
No, please don't alias this. The type is intended to have a documenting purpose. If you alias it things become much less obvious...

paul-michalik commented on src/Frame.cc:
```diff
@@ -277,46 +351,45 @@ void Frame::AssignFeaturesToGrid()
-			}
-		}
-	}
+     if (std::get<ext::tsr_info_opt_t>(_sensor_input)) {
```
Please declare the variables inside the `if` statement: 
```
if (auto tsr_data = std::get<ext::tsr_info_opt_t>(_sensor_input)) {
    // ...
```
Use this approach in every context this is applicable.

paul-michalik commented on src/Frame.cc:
```diff
@@ -277,46 +351,45 @@ void Frame::AssignFeaturesToGrid()
-	
-	if ( (NULL != mpTraficsignGrp) && (true == mpTraficsignGrp->isLoaded))
+
+	if (std::get<ext::tsr_info_opt_t>(_sensor_input))
```
```
if (auto tsr_data = std::get<ext::tsr_info_opt_t>(_sensor_input)) {
    // ...
}
```

paul-michalik commented on src/System.cc:
```diff
@@ -247,6 +247,55 @@ cv::Mat System::TrackRGBD(const cv::Mat &im, const cv::Mat &depthmap, const doub
     return Tcw;
 }
 
+cv::Mat System::TrackMonocular(ext::slam_input_t const& slam_input)
```
In order to reduce code duplication, please forward the standard API call `cv::Mat System::TrackMonocular(const cv::Mat &im, const double &timestamp) ` to this method by creating a tuple:

```
TrackMonocular(std::make_tuple(ext::time_point_t{time_stamp}, ext::image_t{}, ...));
```

paul-michalik commented on src/Tracking.cc:
```diff
@@ -256,13 +254,38 @@ cv::Mat Tracking::GrabImageMonocular(const cv::Mat &im, const double &timestamp)
+    return mCurrentFrame.mTcw.clone();
+}
+
+cv::Mat Tracking::GrabImageMonocular(ext::slam_input_t const& slam_input)
```
Again, can we reduce code duplication by forwarding the standard call to this one?

apattnaik0721013 commented:
@paul-michalik 
1. removed alias
2. remove duplicated code for TrackMonocular()
3. only extra changes i modified along with your comment is the following in Frame.cc
	if (std::get<ext::tsr_info_opt_t>(_sensor_input))
               auto ts_data = std::get<ext::tsr_info_opt_t>(_sensor_input).get();
       Note:  ext::tsr_info_opt_t os  -> is boost:optional variable. 
       that is why , i cant use the following in one line.
       if(auto ts_data = std::get<ext::tsr_info_opt_t>(_sensor_input))

apattnaik0721013 commented:
@paul-michalik 
1. removed Boost::none with the empty constructor
2. use the following code like you proposed
if (auto tsr_data = std::get<ext::tsr_info_opt_t>(_sensor_input))

paul-michalik commented:
@apattnaik0721013 have you re-written the whole history of the branch!? I can't associate a single commit...
